### PR TITLE
Fix nid assignment in ASN1_STRING_TABLE_add

### LINKED ISSUE #3923: ASN1_STRING_TABLE_add doesn't set the nid for new items
If the nid is not found in the predefined table (ASN1_STRING_TABLE_add) the new element is created without setting it's nid field when using ASN1_STRING_TABLE_add. Because of this ASN1_STRING_TABLE_get returns NULL for all new added values.

Here is the suggested fix:
```
int ASN1_STRING_TABLE_add(int nid,
                          long minsize, long maxsize, unsigned long mask,
                          unsigned long flags)
{
    ASN1_STRING_TABLE *tmp;
    tmp = stable_get(nid);
    if (!tmp) {
        ASN1err(ASN1_F_ASN1_STRING_TABLE_ADD, ERR_R_MALLOC_FAILURE);
        return 0;
    }
    tmp->nid = nid;   //required fix
    if (minsize >= 0)
        tmp->minsize = minsize;
    if (maxsize >= 0)
        tmp->maxsize = maxsize;
    if (mask)
        tmp->mask = mask;
    if (flags)
        tmp->flags = STABLE_FLAGS_MALLOC | flags;
    return 1;
}
```

### PR DESCRIPTION
**Pull request by lolyonok**
Fixes #3923

### CODE REVIEW

File: crypto/asn1/a_strnid.c [hunk 1]
```diff
@@ -184,6 +184,7 @@ int ASN1_STRING_TABLE_add(int nid,
         ASN1err(ASN1_F_ASN1_STRING_TABLE_ADD, ERR_R_MALLOC_FAILURE);
         return 0;
     }
+    tmp->nid = nid;
```
Comment (InfoHunter) [inline crypto/asn1/a_strnid.c]: Should it be more proper to put this into `stable_get` function with:

```
    if (tmp) {
        rv->nid = tmp->nid;
        rv->minsize = tmp->minsize;
        rv->maxsize = tmp->maxsize;
        rv->mask = tmp->mask;
        rv->flags = tmp->flags | STABLE_FLAGS_MALLOC;
    } else {
        rv->minsize = -1;
        rv->maxsize = -1;
        rv->flags = STABLE_FLAGS_MALLOC;
    }
```

**Comment by lolyonok** [comment]
I was thinking about it as well. Moved to stable_get.

**Comment by richsalz** [comment]
Please sign our cla or edit your commit message to put "CLA: trivial" *in the text of the commit.*
This looks trivial to me, thanks!

**Comment by InfoHunter** [comment]
The year in the following line in this file might be also updated to '1999-2017', would you please change that incidentally? @lolyonok : )

> Copyright 1999-2016 The OpenSSL Project Authors. All Rights Reserved.

**Comment by InfoHunter** [comment]
Ping @openssl to review?

**Comment by richsalz** [comment]
We need the "CLA: trivial"

**Comment by kroeckx** [comment]
@lolyonok: We need you to either say this is trivial or sign the CLA.

**Comment by lolyonok** [comment]
I've added "CLA: trivial" into the commit message after you asked it for the first time. All the checks passed with it. Should I do something more?

**Comment by richsalz** [comment]
@lolyonok thanks.  our script seems glitchy :(  @levitte any clues?  @kroeckx you want to merge or me?
(Is that a record for most at-refs in a single line? :)

**Comment by richsalz** [comment]
merged, thank you.  @kroeckx should we backport ?

**Comment by kroeckx** [comment]
I understand this is a bug fix, so we should backport it.

**Comment by richsalz** [comment]
picked to 1.1.0; 1.0.2 code is different (no stable_get; and the nid value is copied in the add routine) so not picked there.

**Pull request closed without merging**